server/tso/allocator_manager.go

```diff
@@ -0,0 +1,168 @@
+type AllocatorManager struct {
+	sync.RWMutex
+	// There are two kinds of TSO Allocators:
+	//   1. Global TSO Allocator, as a global single point to allocate
+	//      TSO for global transactions, such as cross-region cases.
+	//   2. Local TSO Allocator, servers for DC-level transactions.
+	// dc-location/global (string) -> TSO Allocator
+	allocatorGroups map[string]*allocatorGroup
+	// etcd and its client
+	etcd   *embed.Etcd
+	client *clientv3.Client
+	// tso config
+	rootPath      string
+	saveInterval  time.Duration
+	maxResetTSGap func() time.Duration
+}
+
+// NewAllocatorManager creates a new TSO Allocator Manager.
+func NewAllocatorManager(serverCtx context.Context, etcd *embed.Etcd, client *clientv3.Client, rootPath string, saveInterval time.Duration, maxResetTSGap func() time.Duration) *AllocatorManager {
+	allocatorManager := &AllocatorManager{
+		allocatorGroups: make(map[string]*allocatorGroup),
+		etcd:            etcd,
+		client:          client,
+		rootPath:        path.Join(rootPath, "allocator"),
+		saveInterval:    saveInterval,
+		maxResetTSGap:   maxResetTSGap,
+	}
+	// Start the deamon for allocators
+	go allocatorManager.AllocatorDaemon(serverCtx)
+	return allocatorManager
+}
+
+func (am *AllocatorManager) getAllocatorPath(dcLocation string) string {
+	return path.Join(am.rootPath, dcLocation)
+}
+
+// SetUpAllocator is used to set up an allocator, which will initialize the allocator and put it into allocator daemon.
+func (am *AllocatorManager) SetUpAllocator(ctx context.Context, cancel context.CancelFunc, dcLocation string, leadership *election.Leadership) error {
+	am.Lock()
+	defer am.Unlock()
+	switch dcLocation {
+	case "global":
+		am.allocatorGroups[dcLocation] = &allocatorGroup{
+			ctx:        ctx,
+			cancel:     cancel,
+			leadership: leadership,
+			allocator:  NewGlobalTSOAllocator(leadership, am.getAllocatorPath(dcLocation), am.saveInterval, am.maxResetTSGap),
+		}
+		if err := am.allocatorGroups[dcLocation].allocator.Initialize(); err != nil {
+			return errs.ErrSetAllocator.FastGenByArgs(err)
+		}
+		am.allocatorGroups[dcLocation].isInitialized = true
+	default:
+		// Todo: set up a Local TSO Allocator
+	}
+	return nil
+}
+
+// GetAllocator get the allocator by dc-location.
+func (am *AllocatorManager) GetAllocator(dcLocation string) (Allocator, error) {
+	am.RLock()
+	defer am.RUnlock()
+	allocatorGroup, exist := am.allocatorGroups[dcLocation]
+	if !exist {
+		return nil, errs.ErrGetAllocator.FastGenByArgs(fmt.Sprintf("%s allocator not found", dcLocation))
+	}
+	return allocatorGroup.allocator, nil
+}
+
+// AllocatorDaemon is used to update every allocator's TSO.
+func (am *AllocatorManager) AllocatorDaemon(serverCtx context.Context) {
+	tsTicker := time.NewTicker(UpdateTimestampStep)
+	defer tsTicker.Stop()
+
+	for {
+		select {
+		case <-tsTicker.C:
+			am.RLock()
+			for dcLocation, allocatorGroup := range am.allocatorGroups {
```

Comment:
maybe better to handle different allocators in separated goroutines. When an allocator gets stuck, it has a smaller impact on others.

Reply:
Ok, let's do it this way.